File: libjava/classpath/java/lang/Void.java
/* Void.class - defines void.class
   Copyright (C) 1998, 1999, 2001, 2002, 2004 Free Software Foundation, Inc.

This file is part of GNU Classpath.

GNU Classpath is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

GNU Classpath is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with GNU Classpath; see the file COPYING.  If not, write to the
Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301 USA.

Linking this library statically or dynamically with other modules is
making a combined work based on this library.  Thus, the terms and
conditions of the GNU General Public License cover the whole
combination.

As a special exception, the copyright holders of this library give you
permission to link this library with independent modules to produce an
executable, regardless of the license terms of these independent
modules, and to copy and distribute the resulting executable under
terms of your choice, provided that you also meet, for each linked
independent module, the terms and conditions of the license of that
module.  An independent module is a module which is not derived from
or based on this library.  If you modify this library, you may extend
this exception to your version of the library, but you are not
obligated to do so.  If you do not wish to do so, delete this
exception statement from your version. */

package java.lang;


/**
 * Void is a placeholder class so that the variable <code>Void.TYPE</code>
 * (also available as <code>void.class</code>) can be supported for
 * reflection return types.
 *
 * <p>This class could be Serializable, but that is up to Sun.</p>
 *
 * @author Paul Fisher
 * @author John Keiser
 * @author Eric Blake (ebb9@email.byu.edu)
 * @since 1.1
 * @status updated to 1.5
 */
public final class Void
{
  /**
   * The return type <code>void</code> is represented by this
   * <code>Class</code> object.
   */
  public static final Class<Void> TYPE = (Class<Void>) VMClassLoader.getPrimitiveClass('V');

  /**
   * Void is non-instantiable.
   */
  private Void()
  {
  }
}

File: libstdc++-v3/testsuite/20_util/owner_less/void.cc
// Copyright (C) 2016-2021 Free Software Foundation, Inc.
//
// This file is part of the GNU ISO C++ Library.  This library is free
// software; you can redistribute it and/or modify it under the
// terms of the GNU General Public License as published by the
// Free Software Foundation; either version 3, or (at your option)
// any later version.

// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License along
// with this library; see the file COPYING3.  If not see
// <http://www.gnu.org/licenses/>.

// { dg-do compile { target c++11 } }

#include <memory>

#if __cplusplus >= 201402L
// The feature-test macro is only defined for C++14 and later.
# if __cpp_lib_transparent_operators < 201510
#  error "__cpp_lib_transparent_operators < 201510"
# endif
#endif

void
test01()
{
  using namespace std;

  static_assert(is_same<owner_less<>, owner_less<void>>::value,
                "owner_less<> uses void specialization");

  shared_ptr<int> sp1;
  shared_ptr<void> sp2;
  shared_ptr<long> sp3;
  weak_ptr<int> wp1;

  owner_less<> cmp;
  cmp(sp1, sp2);
  cmp(sp1, wp1);
  cmp(sp1, sp3);
  cmp(wp1, sp1);
  cmp(wp1, wp1);
}
